text/0055-remove-unspecified.md

```diff
@@ -0,0 +1,199 @@
+This model can feel awkward in cases where there is not a clear principal and/or resource for an action.
+For example, consider a "createFile" action that can be used by any `User`, as encoded in the following policy:
+
+```
+permit(principal is User, action == Action::"createFile", resource);
+```
+
+When making an authorization request for this action, the principal should be a `User`, and the context can store information about the file to create, but it is unclear what the resource should be.
+In fact, given the policy above, the resource _doesn't matter_.
+In this case, a Cedar user may want to make the authorization request with a “dummy” resource; this is what Cedar’s unspecified entities are for.
+
+As of Cedar v3.1, the only way to “construct” an unspecified entity is by passing `None` as a component when building a [`Request`](https://docs.rs/cedar-policy/3.0.1/cedar_policy/struct.Request.html).
+So a request for this action might look like \<`Some(User::"alice")`, `Some(Action::"createFile")`, `None`, ...\> (using pseudo-syntax).
+
+In the schema, omitting the relevant `appliesTo` field says that an action applies to an unspecified entity.
+For example, the following schemas specify that the "createFile" action applies (only) to an unspecified resource.
+
+Cedar schema syntax ([RFC 24](https://github.com/cedar-policy/rfcs/blob/main/text/0024-schema-syntax.md)):
+
+```
+action createFile
+  appliesTo { principal: [User] };
+```
+
+Cedar schema JSON syntax:
+
+```json
+"createFile": {
+    "appliesTo": {
+        "principal": ["User"]
+    }
+}
+```
+
+Now consider a "createAccount" action.
+It's unclear what principal _or_ resource should be used with this action.
+The approach above can be extended to this case too: make the action apply to an unspecified principal _and_ resource.
+But now "unspecified" entities are being used to represent multiple distinct concepts: a file store, (unidentified) users creating an account, and a system that manages accounts.
+
+This RFC proposes that users should instead explicitly create these "dummy" entities in an application-specific manner.
+For this example, the user should create entities of three types: `FileSystem`, `UnauthenticatedUser`, and `AccountManager`.
+The first will act as the principal for a "createFile" request, and the second and third will act as the principal and resource for a "createAccount" request.
+
+This is exactly the approach we used in the [TinyTodo application](https://github.com/cedar-policy/cedar-examples/tree/release/3.0.x/tinytodo): we made a special `Application::“TinyTodo”` entity to act as the resource for the “CreateList” and “GetLists” actions.
+This proposal is also in line with [our previous suggestion](https://cedar-policy.slack.com/archives/C0547KH7R19/p1706656288284189) to use an application-specific `Unauthenticated` type to represent "unauthenticated" users, rather than using unspecified entities.
```

Comment:
Re @mwhicks1's comment above (pinning to the text to enable threaded discussion):

> I'm not convinced that `Application::"TinyTodo"` is the same thing as the unspecified entity. Per Alternative A, the unspecified entity should never appear in policies. But `Application::"TinyTodo"` explicitly appears, for a reason. The unspecified entity is useful when principal, resource, etc. are _irrelevant_, and never need to be "looked at" in a policy.

I'm still convinced that it is 🙂 `Application::"TinyTodo"` is an artificial entity with no attributes that is never used in policies conditions. In it's current form, it could have just as well been implemented using an unspecified entity.

Reply:
The reason I think they are different is the motivation for Alternative A: You should not be able to reference the unspecified entity in policies. But that's exactly what we do for `Application::"TinyTodo"`:
```
permit(principal, action in [Action::"CreateList", Action::"GetLists"], resource == Application::"TinyTodo");
```
To use the unspecified entity, we'd have to write the policy thus:
```
permit(principal, action in [Action::"CreateList", Action::"GetLists"], resource);
```
Listing the resource explicitly has at least two advantages:
- It's [useful for indexing](https://docs.cedarpolicy.com/bestpractices/bp-populate-policy-scope.html), which relies on scope constraints.
- In general, we think of `Application::"TinyTodo"` as a list "container", which [is useful to name explicitly in the policy](https://docs.aws.amazon.com/verifiedpermissions/latest/userguide/design-resources-in-containers.html). The reason is, in part, to avoid inadvertently authorizing a request. In this example, a client that mistakenly passes in a resource like `List::"mylist"` would still satisfy the policy, e.g., perhaps thinking that `Action::"GetLists"` is a way to show a list's contents.

Reply:
A followup question: in my (limited) imagination, the only time a user would use an unspecified entity is when they are calling an action with a generic principal (e.g., an unauthenticated user) or an generic resource (e.g., an application, a file system, etc). If you think the resource case is better handled by a custom entity type (like `Application`), then do you still think that unspecified entities are really that useful?

Reply:
> To your second point, using `List::"mylist"` with `Action::"GetLists"` would result in a request validation error,

This is a good point. Of course, you will only get an error if you turn on request validation. Having the `== Application::"TinyTodo"` in the policy makes the situation evident in the policy, rather than in the schema, which I think is nicer. (Indeed, that's one reason we chose to include an `is` operator -- writing `resource is List` makes the policy clearer than simply saying that a particular action only applies to `List` objects in the schema.)

> do you still think that unspecified entities are really that useful?

I don't think they are that useful. :) To me, their best use is for actions that simply do not need a principal and/or resource at all, but the structure of Cedar requests and policies obligates a client to provide one. The "unspecified" resource acts as that stand-in.

Reply:
I also think `Application::"TinyTodo"` should be viewed as an unspecified entity. My reasoning is that a policy store should only contain policies for the application in question, so saying 'permit ... when I'm in Application::"TinyTodo"' doesn't add information.

I don't find the objection that users may apply `Action::"GetLists"` to a `List` convincing. When writing code, I'd expect the author to refer to the schema. Assuming they misunderstand `Action::"GetLists"`, they should reach some point like:
```
fn GetLists(user: User, list: List) -> Vec<ListItem> {
  let req = build_req_get_lists(user, list);
  if authorizer.is_authorized(req) {
    return Error::Unauthorized();
  }
  ...
```
The author needs to build a request given `user` and `list`. To decide how they will almost certainly look at the schema, and it will quickly become clear they misunderstand what `Action::"GetLists"` does.
I suspect adding `Application::"TinyTodo"` will actually make this more confusing. They need to understand why this entity was created and that there is no corresponding argument for them to pass. Using unspecified, it becomes clear that there's no `List` argument needed in the above code.

I do think `Action::"GetLists"` is poorly named, but that's a separate problem.